Source/DafnyLanguageServer/Workspace/DocumentDatabase.cs

```diff
@@ -99,65 +109,107 @@ ISymbolTableRelocator symbolTableRelocator
+    private async Task<DafnyDocument> ApplyChangesAsync(Task<DafnyDocument> oldDocumentTask, DidChangeTextDocumentParams documentChange, CancellationToken cancellationToken) {
+#pragma warning disable VSTHRD003
+      var oldDocument = await oldDocumentTask;
+#pragma warning restore VSTHRD003
       // We do not pass the cancellation token to the text change processor because the text has to be kept in sync with the LSP client.
       var updatedText = textChangeProcessor.ApplyChange(oldDocument.Text, documentChange, CancellationToken.None);
+      var oldVerificationDiagnostics =
+        oldDocument.Errors.GetDiagnostics(oldDocument.Uri).Where(d => d.Code == "Verification").Concat(oldDocument.OldVerificationDiagnostics).ToList();
+      var migratedVerificationDiagnotics =
+        relocator.RelocateDiagnostics(oldVerificationDiagnostics, documentChange, CancellationToken.None);
       try {
-        var newDocument = await LoadAndVerifyAsync(updatedText, VerifyOnChange, cancellationToken);
+        var newDocument = await documentLoader.LoadAsync(updatedText, cancellationToken);
         if (newDocument.SymbolTable.Resolved) {
-          return newDocument;
+          return newDocument with {
+            OldVerificationDiagnostics = migratedVerificationDiagnotics
+          };
         }
         // The document loader failed to create a new symbol table. Since we'd still like to provide
         // features such as code completion and lookup, we re-locate the previously resolved symbols
         // according to the change.
         return newDocument with {
-          SymbolTable = symbolTableRelocator.Relocate(oldDocument.SymbolTable, documentChange, CancellationToken.None)
+          SymbolTable = relocator.RelocateSymbols(oldDocument.SymbolTable, documentChange, CancellationToken.None),
+          OldVerificationDiagnostics = migratedVerificationDiagnotics
         };
       } catch (OperationCanceledException) {
         // The document load was canceled before it could complete. We migrate the document
         // to re-locate symbols that were resolved previously.
         logger.LogTrace("document loading canceled, applying migration");
         return oldDocument with {
           Text = updatedText,
-          SymbolTable = symbolTableRelocator.Relocate(oldDocument.SymbolTable, documentChange, CancellationToken.None),
+          SymbolTable = relocator.RelocateSymbols(oldDocument.SymbolTable, documentChange, CancellationToken.None),
           SerializedCounterExamples = null,
-          LoadCanceled = true
+          LoadCanceled = true,
+          OldVerificationDiagnostics = migratedVerificationDiagnotics
         };
       }
     }
 
-    public Task<DafnyDocument> SaveDocumentAsync(TextDocumentIdentifier documentId) {
+    public IObservable<DafnyDocument> SaveDocument(TextDocumentIdentifier documentId) {
```

Comment:
This change will conflict with @atomb 's changes. @atomb can you comment on this version?

Reply:
I think the conflicts should be fine to resolve :-)